core/blockchain.go

```diff
@@ -1214,7 +1223,7 @@ func (bc *BlockChain) writeBlockWithState(block *types.Block, receipts []*types.
 	triedb := bc.stateCache.TrieDB()
 
 	// If we're running an archive node, always flush
-	if bc.cacheConfig.TrieDirtyDisabled {
+	if mode := atomic.LoadInt32(&bc.gcmode); mode == archiveMode {
 		return triedb.Commit(root, false, nil)
 	} else {
```

Comment:
The upper clause `return`s, let's get rid of the `else` and unindent the next clause

Reply:
And after that, maybe 
```
	if current := block.NumberU64(); current <= TriesInMemory {
		return nil
	}
```
And we can un-indent some more